### DISCUSSION TITLE
Error while extracting Frame Position in OpenVSLAM

### DISCUSSION THREAD

**Discussion by Udaykiran87** [stella-cv/stella_vslam#100] [Q&A] [answered]
Hello Community,

I was trying to extract the frame position or get localized frame position as instructed below:
**https://github.com/xdspacelab/openvslam/issues/292**

But i get below error:
**/openvslam/example/run_video_slam.cc:74:50: error: invalid use of incomplete type 'using element_type = class openvslam::publish::map_publisher {aka class openvslam::publish::map_publisher}'
                 auto camera_pose = map_publisher_->get_current_cam_pose().inverse().eval();**

Note:
Unfortunately I do not know c++.

Modified Code snippet is as below (from the line '// extract frame position'):
file name: **run_video_slam.cc**. Same modification has been done in **run_video_localization.cc**

```
            if (!frame.empty() && (num_frame % frame_skip == 0)) {
                // input the current frame and estimate the camera pose
                SLAM.feed_monocular_frame(frame, timestamp, mask);
                // extract frame position
                auto map_publisher_ = SLAM.get_map_publisher();
                auto camera_pose = map_publisher_->get_current_cam_pose().inverse().eval();
                // log message to file
                try{
                   auto pos_logger = spdlog::basic_logger_mt("frame position logger","mapping_pos_logs.txt");
                   pos_logger->info(camera_pose);
                } catch (const spdlog::spdlog_ex &ex)   {
                   std::cout << ex.what() << std::endl;   
                }
            }

```
**Detailed error log:**

Scanning dependencies of target run_video_localization
[ 95%] Building CXX object example/CMakeFiles/run_video_localization.dir/run_video_localization.cc.o
[ 95%] Built target run_camera_slam
[ 96%] Building CXX object example/CMakeFiles/run_tum_rgbd_slam.dir/util/tum_rgbd_util.cc.o
[ 97%] Linking CXX executable ../run_tum_rgbd_slam
**/openvslam/example/run_video_slam.cc: In lambda function:
/openvslam/example/run_video_slam.cc:74:50: error: invalid use of incomplete type 'using element_type = class openvslam::publish::map_publisher {aka class openvslam::publish::map_publisher}'
                 auto camera_pose = map_publisher_->get_current_cam_pose().inverse().eval();
                                                  ^~
In file included from /openvslam/src/socket_publisher/publisher.h:4:0,
                 from /openvslam/example/run_video_slam.cc:4:
/openvslam/src/socket_publisher/data_serializer.h:23:7: note: forward declaration of 'using element_type = class openvslam::publish::map_publisher {aka class openvslam::publish::map_publisher}'
 class map_publisher;
       ^~~~~~~~~~~~~**
[ 97%] Built target run_tum_rgbd_slam
****/openvslam/example/run_video_localization.cc: In lambda function:
/openvslam/example/run_video_localization.cc:82:50: error: invalid use of incomplete type 'using element_type = class openvslam::publish::map_publisher {aka class openvslam::publish::map_publisher}'
                 auto camera_pose = map_publisher_->get_current_cam_pose().inverse().eval();**
                                                  ^~
In file included from /openvslam/src/socket_publisher/publisher.h:4:0,
                 from /openvslam/example/run_video_localization.cc:4:
/openvslam/src/socket_publisher/data_serializer.h:23:7: note: forward declaration of 'using element_type = class openvslam::publish::map_publisher {aka class openvslam::publish::map_publisher}'
 class map_publisher;
       ^~~~~~~~~~~~~
make[2]: *** [example/CMakeFiles/run_video_slam.dir/run_video_slam.cc.o] Error 1
example/CMakeFiles/run_video_slam.dir/build.make:62: recipe for target 'example/CMakeFiles/run_video_slam.dir/run_video_slam.cc.o' failed
CMakeFiles/Makefile2:927: recipe for target 'example/CMakeFiles/run_video_slam.dir/all' failed
make[1]: *** [example/CMakeFiles/run_video_slam.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
make[2]: *** [example/CMakeFiles/run_video_localization.dir/run_video_localization.cc.o] Error 1
make[1]: *** [example/CMakeFiles/run_video_localization.dir/all] Error 2
example/CMakeFiles/run_video_localization.dir/build.make:62: recipe for target 'example/CMakeFiles/run_video_localization.dir/run_video_localization.cc.o' failed
CMakeFiles/Makefile2:966: recipe for target 'example/CMakeFiles/run_video_localization.dir/all' failed**
make: *** [all] Error 2
Makefile:129: recipe for target 'all' failed
The command '/bin/sh -c set -x &&   mkdir -p build &&   cd build &&   cmake     -DUSE_SOCKET_PUBLISHER=ON     -DBUILD_EXAMPLES=ON     -DUSE_STACK_TRACE_LOGGER=ON     .. &&   make -j${NUM_THREADS} &&   rm -rf CMakeCache.txt CMakeFiles Makefile cmake_install.cmake example src &&   chmod -R 777 ./*' returned a non-zero code: 2

Could you please help me to resolve error.
Is there any pythonic way of doing it or any python interface to call c++ for OpenVSLAM?

Thanks and Regards,
Udaykiran Patnaik

**Comment by Udaykiran87** [comment, marked as answer]
Hello,
I solved the issue. Since i do not have much knowledge on cpp, so took time to debug.
I had missed below includes, hence the error occured.

**#include "openvslam/publish/map_publisher.h"**

But as i asked earlier is there any pythonic way that i can achieve this task? or any python interface to call c++ for OpenVSLAM?

Thanks and Regards,
Udaykiran Patnaik.

**Reply by ymd-stella** [reply]
I don't know what you are going to use it for, but if you need to get values in real time, I recommend using https://github.com/OpenVSLAM-Community/openvslam_ros. With rclpy, you can subscribe to Pose and create your own application.

**Reply by Udaykiran87** [reply]
@ymd-stella ,
Thank you so much. I will refer to it.

Thanks and Regards,
Udaykiran Patnaik.
